File: extract_links.js

const fs = require('fs');
const path = require('path');

const txt_dir = 'txt';
const categories_mapping = {
    "AI Services": "ai-services.html",
    "Colors": "colors.html",
    "Developer": "developer.html",
    "Help Online Resources": "help.html",
    "Illustrations": "illustrations.html",
    "MUSIC DESIGN": "music.html",
    "Software": "software.html",
    "UX": "uiux.html",
    "Video": "photos-video.html",
    "Web Design": "web-design.html"
};

const extract_links = (content) => {
    // Regex for URLs ending with any TLD (2+ letters)
    const tld_pattern = /\b(?:[a-zA-Z0-9-]+\.)+[a-z]{2,10}(?:\/[^\s]*)?\b/gi;
    
    // Clean up content
    content = content.replace(/Figma File[\s\S]*/gi, '');
    content = content.replace(/Des igned and Created by[\s\S]*/gi, '');
    
    const links = [];
    const matches = Array.from(content.matchAll(tld_pattern));
    
    let lastIndex = 0;
    matches.forEach(match => {
        const url = match[0].trim();
        const start = match.index;
        
        // Take content between URLs
        let desc = content.slice(lastIndex, start).trim();
        
        // If desc contains newlines, take the last non-empty line
        if (desc.includes('\n')) {
            const lines = desc.split('\n');
            desc = lines[lines.length - 1].trim() || lines[lines.length - 2]?.trim() || '';
        }
        
        if (desc && url) {
            links.push({ name: desc, url: url });
        }
        lastIndex = start + url.length;
    });
            
    return links;
};

const all_data = {};

const files = fs.readdirSync(txt_dir);
files.forEach(filename => {
    if (!filename.endsWith(".txt")) return;
    
    // Normalize base category name
    let base_name = filename.replace('.txt', '').replace(/\s*\d*$/, '');
    if (base_name === "Web Design-1") base_name = "Web Design";
    if (base_name === "Help Online Resources") base_name = "Help Online Resources";
    
    let match_cat = null;
    Object.keys(categories_mapping).forEach(cat => {
        if (base_name.toLowerCase().startsWith(cat.toLowerCase()) || cat.toLowerCase().startsWith(base_name.toLowerCase())) {
            match_cat = cat;
        }
    });
            
    if (!match_cat) {
        // Find by partial match if needed
        console.log(`No match for ${filename}`);
        return;
    }
        
    const content = fs.readFileSync(path.join(txt_dir, filename), 'utf8');
    const extracted = extract_links(content);
    if (!all_data[match_cat]) all_data[match_cat] = [];
    all_data[match_cat] = all_data[match_cat].concat(extracted);
});

// Deduplicate by URL
Object.keys(all_data).forEach(cat => {
    const seen = new Set();
    const unique = [];
    all_data[cat].forEach(item => {
        if (!seen.has(item.url)) {
            seen.add(item.url);
            unique.push(item);
        }
    });
    all_data[cat] = unique;
});

// Log statistics
let total = 0;
Object.keys(all_data).forEach(cat => {
    console.log(`${cat}: ${all_data[cat].length} links`);
    total += all_data[cat].length;
});
console.log(`\nTotal unique links: ${total}`);

fs.writeFileSync('extracted_links.json', JSON.stringify(all_data, null, 2), 'utf8');
console.log('\nJSON saved to extracted_links.json');

File: extract_links.py

import os
import re
import json

txt_dir = r"e:\my\apps - for playstore\generalist-bible\txt"
categories_mapping = {
    "AI Services": "ai-services.html",
    "Colors": "colors.html",
    "Developer": "developer.html",
    "Help Online Resources": "help.html",
    "Illustrations": "illustrations.html",
    "MUSIC DESIGN": "music.html",
    "Software": "software.html",
    "UX": "uiux.html",
    "Video": "photos-video.html",
    "Web Design": "web-design.html"
}

def extract_links(content):
    # This is a bit tricky. We want to find URLs and the text before them.
    # URLs often end with .com, .io, .net, etc.
    # We'll use a regex for common tlds.
    tld_pattern = r'\b(?:[a-zA-Z0-9-]+\.)+(?:com|io|net|org|tt|fm|xyz|club|sh|me|site|co|app|ru|in|wtf|gallery|id|dev|energy|cc|pro|media|ai|app|so|build|ly|is|to|cc|me|site|plus|top|tech|online|best|info)\b[^\s]*'
    
    links = []
    
    # Clean up content (remove markers like "Figma File" or footer)
    content = re.sub(r'Figma File.*', '', content, flags=re.DOTALL)
    content = re.sub(r'Des igned and Created by.*', '', content, flags=re.DOTALL)
    
    # Split by the tld_pattern but keep the matches
    parts = re.split(f'({tld_pattern})', content)
    
    # Pair items: description, url, description, url...
    for i in range(1, len(parts), 2):
        url = parts[i].strip()
        desc = parts[i-1].strip()
        
        # The description might still have some of the previous entry's description.
        # But since we split by url, the desc is what's between URLs.
        # We need to extract only the part relevant to THIS URL.
        # Actually, in the text files, it seems to be: [Description] [URL] [Description] [URL]
        
        # Clean up desc: take the last line if multiple lines, and maybe limit len.
        if '\n' in desc:
            desc = desc.split('\n')[-1].strip()
        
        # If desc is too long, it might be the start of the next one? No, in the txt it's usually short.
        # Example: "Image filters civitai.com" -> desc="Image filters", url="civitai.com"
        
        if desc and url:
            links.append({'name': desc, 'url': url})
            
    return links

all_data = {}

for filename in os.listdir(txt_dir):
    if not filename.endswith(".txt"):
         continue
         
    # Find base category name
    base_name = re.sub(r'\s*\d*$', '', filename.replace('.txt', ''))
    if base_name == "Web Design-1": base_name = "Web Design"
    
    match = None
    for cat in categories_mapping:
        if base_name.startswith(cat) or cat.startswith(base_name):
            match = cat
            break
            
    if not match:
        print(f"No match for {filename}")
        continue
        
    with open(os.path.join(txt_dir, filename), 'r', encoding='utf-8', errors='ignore') as f:
        content = f.read()
        extracted = extract_links(content)
        if match not in all_data:
            all_data[match] = []
        all_data[match].extend(extracted)

# Deduplicate
for cat in all_data:
    seen = set()
    unique = []
    for item in all_data[cat]:
        if item['url'] not in seen:
            seen.add(item['url'])
            unique.append(item)
    all_data[cat] = unique

print(json.dumps(all_data, indent=2))
